model/waterfall.go

```diff
@@ -135,7 +137,109 @@ func getBuildVariantFilterPipeline(ctx context.Context, variants []string, match
 	return pipeline, nil
 }
 
+// GetActiveVersionsByTaskFilters returns limit versions that satisfy a task name or status filter. It also applies any requester and build variant filters.
+// If neither of these filters is specified, use GetActiveWaterfallVersions: it's faster.
+func GetActiveVersionsByTaskFilters(ctx context.Context, projectId string, opts WaterfallOptions, searchOffset int) ([]Version, error) {
+	match := bson.M{
+		task.ProjectKey: projectId,
+		task.RequesterKey: bson.M{
+			"$in": opts.Requesters,
+		},
+		task.ActivatedKey: true,
+	}
+
+	if opts.MaxOrder != 0 && opts.MinOrder != 0 {
+		return nil, errors.New("cannot provide both max and min order options")
+	}
+
+	pagingBackward := opts.MinOrder != 0
+
+	revisionFilter := bson.M{}
+	if pagingBackward {
+		revisionFilter["$lte"] = searchOffset + MaxWaterfallVersionLimit
+		revisionFilter["$gt"] = searchOffset
+	} else {
+		revisionFilter["$gte"] = searchOffset - MaxWaterfallVersionLimit
+		revisionFilter["$lt"] = searchOffset
+	}
+	match[task.RevisionOrderNumberKey] = revisionFilter
+
+	if len(opts.Statuses) > 0 {
+		match[task.DisplayStatusCacheKey] = bson.M{"$in": opts.Statuses}
+	}
+
+	if len(opts.Tasks) > 0 {
+		taskNamesAsRegex := strings.Join(opts.Tasks, "|")
+		match[task.DisplayNameKey] = bson.M{"$regex": taskNamesAsRegex, "$options": "i"}
+	}
+
+	if len(opts.Variants) > 0 {
+		variantsAsRegex := strings.Join(opts.Variants, "|")
+		match["$or"] = []bson.M{
+			{task.BuildVariantKey: bson.M{"$regex": variantsAsRegex, "$options": "i"}},
+			{task.BuildVariantDisplayNameKey: bson.M{"$regex": variantsAsRegex, "$options": "i"}},
+		}
+	}
+
+	pipeline := []bson.M{{"$match": match}}
+
+	pipeline = append(pipeline, bson.M{
+		"$group": bson.M{
+			task.IdKey: "$" + task.VersionKey,
+			// All tasks with the same version key should have the same order number, but $max is the safest way to ensure we can sort from newest to oldest upon grouping.
+			task.RevisionOrderNumberKey: bson.M{
+				"$max": "$" + task.RevisionOrderNumberKey,
+			},
+		},
+	})
+
+	if pagingBackward {
+		// When querying with a $gt param, sort ascending so we can take `limit` versions nearest to the MinOrder param
+		pipeline = append(pipeline, bson.M{"$sort": bson.M{task.RevisionOrderNumberKey: 1}})
+		pipeline = append(pipeline, bson.M{"$limit": opts.Limit})
+		// Then apply an acending sort so these versions are returned in the expected descending order
+		pipeline = append(pipeline, bson.M{"$sort": bson.M{task.RevisionOrderNumberKey: -1}})
+
+	} else {
+		pipeline = append(pipeline, bson.M{"$sort": bson.M{task.RevisionOrderNumberKey: -1}})
+		pipeline = append(pipeline, bson.M{"$limit": opts.Limit})
+
+	}
+
+	versionLookupKey := "version"
```

Comment:
not sure how much of a difference it would make but i imagine you could maybe also run the pipeline at this point, then run [`VersionByIds`](https://github.com/evergreen-ci/evergreen/blob/89695f8f07a107eee07ccc4cf5fdf35070c9da14/model/version_db.go#L64) on the list of version IDs 

(nothing to do im just bringing up alternatives if it sounds like it could help)